mesonbuild/dependencies/python.py

```diff
@@ -271,6 +271,21 @@ def get_windows_link_args(self) -> T.Optional[T.List[str]]:
                         libpath = Path(f'python{vernum}.dll')
                 else:
                     libpath = Path('libs') / f'python{vernum}.lib'
+                    # For a debug build, pyconfig.h forces linking with
+                    # pythonX_d.lib (see meson#10776). This cannot be avoided
+                    # and won't work unless we also have a debug build of
+                    # Python itself. So error out here.
+                    buildtype = self.env.coredata.get_option(mesonlib.OptionKey('buildtype'))
+                    assert isinstance(buildtype, str)
+                    debug = self.env.coredata.get_option(mesonlib.OptionKey('debug'))
+                    is_debug_build = debug or buildtype.startswith('debug')
```

Comment:
I'm not sure the diagnosis is entirely correct. I can build Python extensions with build type set to `debugoptimized` with MSVC without having a debug build of Python. I haven't read all the discussion, but what causes the `_DEBUG` macro to be set? Is it Meson based on the build type or is it based on the value of the `debug` option?

Reply:
From https://learn.microsoft.com/en-us/cpp/c-runtime-library/debug?view=msvc-170: _The compiler defines `_DEBUG` when you specify the `/MTd` or `/MDd` option._

From Meson's builtin options docs:
```
| debug          | `/MDd`         | `/MTd`                |
```

The build type should toggle that same flag (frm [here](https://mesonbuild.com/Builtin-options.html#details-for-buildtype)): _For example, `-Dbuildtype=debugoptimized` is the same as `-Ddebug=true -Doptimization=2` and vice-versa._

If `debugoptimized` doesn't set `_DEBUG` for you, something seems off.

Reply:
> The build type should toggle that same flag (from [here](https://mesonbuild.com/Builtin-options.html#details-for-buildtype)): _For example, `-Dbuildtype=debugoptimized` is the same as `-Ddebug=true -Doptimization=2` and vice-versa._

We already discussed this when we discussed the default build options for meson-python: this part of the documentation is just wrong AFAICT. There are open issues about it and I open one too #11645 with some more details.

Reply:
Okay, then the condition for emitting this warning may need a slight tweak. It's definitely happening for the default settings (which is `buildtype=debug`), so a lot of folks are going to run into it. And it definitely disappears when you set the buildtype to `release`.

So maybe `buildtype.startswith('debug')` should be tweaked to `buildtype == 'debug'`?

I don't really want to get into sorting out those other issues now.

Reply:
> Wouldn't a couple of paragraphs in the documentation for the Meson python module a better solution?

I'd much prefer this warning to be honest. We have the default Meson settings fail with an incredibly obscure link error, and that is tripping people up. I didn't get to the bottom of it when I first opened this PR, and only revisited when someone else pinged me about MSVC not working for a super simple C or Cython extension module.

> Indeed, in meson-build we set `-Db_vscrt=md` to sidestep this issue.

We also set `-Dbuildtype=release`, so this shouldn't matter either way now (assuming you meant `meson-python`). But it doesn't help users who use `meson setup` directly. 

I can add the check for this option too, disabling the warning if the user has explicitly set `b_vscrt=md` or one of the other values that avoids the problem.

Reply:
Yes, I meant meson-python, sorry. I didn't want to suggest to solve the bigger issues as part of this effort. I understand that having the build break with the default settings is at least mildly irritating 😃. However, phrased as it is, the suggestion in the warning is also irritating: why this fine piece of technology does not allow me to have a debug build of my Python extension module unless I procure myself an obscure debug build of Python?

Another issue with the warning is that it is emitted during setup, possibly among quite a bit of other output, and what breaks is the compilation. Thinking about it, I think we can predict exactly when the build is going to break, thus we can turn the warning into an error. If the `b_vscrt` is set to `mdd` or `mtd`, or `b_vscrt` is set to `static_from_buildtype` or `from_buildtype` and `buildtype` is `debug`, and we cannot find `pythonXY_d.lib` we can simply error out.

Reply:
>  why this fine piece of technology does not allow me to have a debug build of my Python extension module unless I procure myself an obscure debug build of Python?

Agreed. I can link to this issue for explanation, or to a CPython bug report (have to open it first). It shouldn't be in the html docs I think, because this should be a temporary situation that we like to get fixed upstream in CPython (which, fingers crossed, is not that hard - it's only deleting some code, so it's mostly getting a core dev to accept the fix).

> and we cannot find `pythonXY_d.lib` we can simply error out.

This would be problematic, for two reasons that I can think of:
1. `pybind11` has a hacky workaround, as I mentioned higher up. So in that case erroring out may be breaking some previously-working code.
2. We're doing these check at configure time in the code for the Python dependency. CPython inserts this link flag only later - where would we put this check.